src/libs/data-utils.js

```diff
@@ -0,0 +1,121 @@
+  }
+
+  async getMetadata() {
+    const { googleProject } = this.props
+    const { uri } = this.state
+    const [bucket, object] = parseUri(uri)
+
+    this.setState({
+      metadata: await Buckets.getObject(bucket, object, googleProject),
+      preview: isFilePreviewable(uri) ? await Buckets.getObjectPreview(bucket, object, googleProject) : undefined
+    })
+  }
+
+  async resolveUri() {
+    const { uri } = this.props
+
+    if (!_.startsWith('gs://', uri)) {
+      this.setState({ uri: await Martha.call(uri) })
+    }
+
+    this.getMetadata()
+  }
+
+  render() {
+    const { uri: originalUri } = this.props
+    const { uri, modalOpen, metadata, preview } = this.state
+
+    return h(Fragment, [
+      link({
+        onClick: () => {
+          this.resolveUri()
+          this.setState({ modalOpen: true })
+        }
+      }, _.startsWith('gs://', originalUri) ? _.last(originalUri.split('/')) : originalUri),
+      modalOpen && h(Modal, {
+        onDismiss: () => this.setState({ modalOpen: false }),
+        title: 'File Details',
+        showCancel: false,
+        okButton: 'Done'
+      },
+      [
+        div({ style: { width: 700 } },
+          Utils.cond([
+            uri, () => {
+              const fileName = _.last(uri.split('/'))
+
+              return [
+                styles.cell([`Filename: ${fileName}`]),
+                h(Fragment, !metadata ? ['Loading metadata...', spinner()] :
+                  [
+                    styles.cell(isFilePreviewable(fileName) ? [
+                      'Preview',
+                      preview ? pre({}, [preview]) :
+                        div({}, ['Loading preview...', spinner()])
+                    ] : ['File can\'t be previewed.']),
+                    styles.cell([`File size: ${filesize(metadata.size)}`]),
+                    styles.cell([
+                      link({
+                        target: 'blank',
+                        href: `https://accounts.google.com/AccountChooser?continue=https://console.cloud.google.com/storage/browser/${parseUri(
+                          uri)[0]}`
+                      }, ['View this file in the Google Cloud Storage Browser'])
+                    ]),
+                    styles.cell([buttonPrimary({}, [`Download for ${formatPriceInDollars(metadata.estimatedCostUSD)}`])]),
+                    styles.cell(['gsutil download command: ', code({}, [`gsutil cp ${uri} .`])]),
+                    h(Collapse, { title: 'More Information', defaultHidden: true }, [
+                      styles.cell([`Created: ${Utils.makePrettyDate(metadata.timeCreated)}`]),
+                      styles.cell([`Updated: ${Utils.makePrettyDate(metadata.updated)}`]),
+                      styles.cell([`md5: ${metadata.md5Hash}`])
+                    ])
+                  ]
+                )
+              ]
+            }
+          ], () => ['Resolving DOS uri...', spinner()])
```

Comment:
TOL: Would it be simpler to unify the spinners here, or do these operations take long enough that it's helpful to give intermediate feedback/status?